src/hotspot/share/classfile/stringTable.hpp

```diff
@@ -56,6 +56,38 @@ class StringTable : AllStatic {
   static double get_load_factor();
   static double get_dead_factor(size_t num_dead);
 
+  // typedef enum { obj_str, unicode_str, symbol_str, utf8_str } StringType;
+  enum class StringType {
+    OopStr, UnicodeStr, SymbolStr, UTF8Str
+  };
+
+  struct StringWrapperInternal {
+    union {
+      const Handle oop_str;
+      const jchar* unicode_str;
+      const Symbol* symbol_str;
+      const char* utf8_str;
+    };
+    const StringType type;
+    const size_t length;
+
+    StringWrapperInternal(const Handle oop_str, const size_t length)     : oop_str(oop_str),         type(StringType::OopStr), length(length)     {}
+    StringWrapperInternal(const jchar* unicode_str, const size_t length) : unicode_str(unicode_str), type(StringType::UnicodeStr), length(length) {}
+    StringWrapperInternal(const Symbol* symbol_str, const size_t length) : symbol_str(symbol_str),   type(StringType::SymbolStr), length(length)  {}
+    StringWrapperInternal(const char* utf8_str, const size_t length)     : utf8_str(utf8_str),       type(StringType::UTF8Str), length(length)    {}
+  };
+
+public:
+  typedef struct StringWrapperInternal StringWrapper;
+
+  static bool wrapped_string_equals(oop java_string, StringWrapper wrapped_str, int _ = 0);
```

Comment:
One thing I can't quite connect the dots on here, is what gets passed as the unused length when this is called from the table code?

Reply:
Inside the hash table, the equals function is called from `lookup`, where the caller passes the length as a parameter. In our case, this is in `lookup_shared`:

https://github.com/openjdk/jdk/blob/4a78433cd33530186b7638a6aefd0e73ec50a3d5/src/hotspot/share/classfile/stringTable.cpp?plain=1#L904-L908

I added an explaining comment as to why 0 is passed instead of the actual length.